### ISSUE TITLE
Just wondering

### ISSUE THREAD

**Issue by patrickjquinn** [karawitan/rust-on-bb10#1]
How far did you get with building rust for BB? I havent tried the setup script set, how far does it get?

**Comment by karawitan** [comment]
Not very far .. After the configure step, "x.py" tries to download a file that leads to a 404 ... 
Very strange .. Tried the 3 latest versions of Rust tonight with no luck

I'll copy paste the steps here, done through a Docker container with python27 as build dependency

<code>
root@11a9c8326913:/src/berrymuch/work/rustc-1.43.1-src# . /src/bbndk/bbndk-env_10_3_1_995.sh 

root@11a9c8326913:/src/berrymuch/work/rustc-1.43.1-src# ./configure --host=arm-unknown-nto-qnx8.0.0eabi --build=i686-linux-gnu --target=arm-unknown-nto-qnx8.0.0eabi
configure: 
configure: build.host           := ['']
configure: install.prefix       := 
configure: build.build          := 
configure: build.target         := ['']
configure: build.configure-args := ['--host=', '--build=', '--target=', '--prefix ...
configure: 
configure: writing `config.toml` in current directory
configure: 
configure: run `python /src/berrymuch/work/rustc-1.43.1-src/x.py --help`
configure: 
root@11a9c8326913:/src/berrymuch/work/rustc-1.43.1-src# 

root@11a9c8326913:/src/berrymuch/work/rustc-1.43.1-src# python2.7 /src/berrymuch/work/rustc-1.43.1-src/x.py --help
info: Downloading and building bootstrap before processing --help
      command. See src/bootstrap/README.md for help with common
      commands.
curl: (22) The requested URL returned error: 404 

spurious failure, trying again
curl: (22) The requested URL returned error: 404 

spurious failure, trying again
curl: (22) The requested URL returned error: 404 

spurious failure, trying again
curl: (22) The requested URL returned error: 404 

spurious failure, trying again
curl: (22) The requested URL returned error: 404 
failed to run: curl -s -y 30 -Y 10 --connect-timeout 30 --retry 3 -Sf -o /tmp/tmpLg0vCL.sha256 https://static.rust-lang.org/dist/2020-03-12/rust-std-1.42.0-''.tar.gz.sha256
root@11a9c8326913:/src/berrymuch/work/rustc-1.43.1-src# 
</code>

**Comment by karawitan** [comment]
404 is at step "download_stage0" in src/bootstrap/bootstrap.py:
<code>

    def download_stage0(self):
        """Fetch the build system for Rust, written in Rust

</code>

The files below https://static.rust-lang.org/dist cannot be browsed easily as far as I tried ..

**Comment by karawitan** [comment]
If you want, we may try to go further.
Do you have Docker installed somewhere ?

**Comment by patrickjquinn** [comment]
I do indeed, happy to try this together with you, is there a dockerfile I can use to configure the container? It might in the repo but I’m reply via email right now 😂

**Comment by karawitan** [comment]
Yep, you may just type "make shell" from the root of the berrymuch/ repo.
This will pull yamsergey/bb10-ndk:0.5.1 docker image, apply some uid tweakings (to have same uid in the container as in the host), then tag the new image as "yamsergey/bb10-ndk:0.5.2" and open a shell inside.

If you want to build the docker image by yourself, a simple "docker build ." will do the trick

**Comment by patrickjquinn** [comment]
The fact you've automated the process within the repo is very welcome, trying this now :) awesome work.

**Comment by karawitan** [comment]
all credits goes to @yamsergey

**Comment by extrowerk** [comment]
> https://static.rust-lang.org/dist/2020-03-12/rust-std-1.42.0-''.tar.gz.sha256

This URL looks incomplete. Maybe an uninitialized variable?

**Comment by patrickjquinn** [comment]
getting into this now (v busy week!), I think extrowerk is right, looking at patching that now. And then will give build another shot. 

Having rust available means I can get librespot built which means i'm a man on a mission. also good experience porting other stuff i'll need going forward :)

**Comment by extrowerk** [comment]
I would like really compile some small cli utilities for my passport, can you guys give me a hint how to set things up correctly? I have the BB SDK and compiler set installed, but never messed with Blackberry development before, so it is unknown region to me.
I got real academical full licensed Momentics suite too, cause why not.

**Comment by patrickjquinn** [comment]
A. Check out the post above, download Docker and install it (also start the service if you're running on Mac or Windows) and then after cloning the berrymuch repo, run make shell and then ssh into your docker image after running it. The add your cli tool that you're hoping to build to the build.sh and you should be go to build from there. 

B. Wow, I wasnt aware there was another version of momentics. What are the benefits of the academic version?

**Comment by extrowerk** [comment]
Momentics originally was made for the QNX operating system.The Blackberry phones running QNX, so Blackberry made a slimmed down version which is freely available for the app developers.

The full Momentics is for embedded systems, it support ARM7+, x86, x86_64, PPC, etc. You basically write your software in Momentics, sets up the deploying settings, and it generates a firmware package with the OS together for your embedded device, flashes it, starts it, debugs it, etc.

QNX is propertiary and non-free software, the flagship product of Blackberry. QNX is still in development (current one is version 7, the phones used QNX6.x branch AFAIK) and available in binary and source form under contracts and NDAs. As mortal human you have only 2 way to mess with it: ask them for academical license or just join to the automotive industry as embedded programmer for infotrainment appliances.

Thanks for the help, btw.

**Comment by extrowerk** [comment]
> A. Check out the post above, download Docker and install it (also start the service if you're running on Mac or Windows) and then after cloning the berrymuch repo, run make shell and then ssh into your docker image after running it. The add your cli tool that you're hoping to build to the build.sh and you should be go to build from there.

Did everything, it spawns a subshell, but the gcc -v command says it is gcc 5.4.0 configured for Ubuntu 16.04.12, and targets i686-linux-gnu, which i belive is not correct.
I sourced the bbndk-env script already.

**Comment by patrickjquinn** [comment]
Well thats weird, its telling you that when you've SSH'd into the docker image? Perhaps its an issue of paths not being aliased properly?

**Comment by extrowerk** [comment]
Will check trough ssh, but i expected it to work directly without the need of ssh.

**Comment by patrickjquinn** [comment]
From my experience it’s easier to do the build from within the container as everything is already linked up and paths set etc. You can of course pass stuff to the containers but I’ve had greater levels of success doing it this way 🙂

**Comment by karawitan** [comment]
/usr/bin/gcc * is not the cross compiler *
the cross compiler is part of the bb ndk below /root/bbndk
/root/bbndk/host_10_3_1_12/linux/x86/usr/bin/arm-unknown-nto-qnx8.0.0eabi-gcc-4.6.3 -v

**Comment by extrowerk** [comment]
> /usr/bin/gcc * is not the cross compiler *

I know, but ‘make shell’ spawned a subshell, and i expected it to work seamlessly. This is my first time with docker.

**Comment by karawitan** [comment]
oki :)  ssh to the container is not an option, as it would mean adding an ssh service, and many useless things to the Docker setup.
"make shell" is the way :)
in the Makefile, you may notice some -v option lines.
They are "bind mounts", ie: volumes from mounted from the host to the container.
For example, the berrymuch/ directory (git repo), in which the Dockerfile is, will be available via /berrymuch, thanks to the following bind mount in the Makefile;

-v "${PWD}":/berrymuch

Usuallly the docker container is running as root.
This was confusing to me, as the files build were owned by root, breaking the permissions of the "host user".

That's why I added the .useradd target in the Makefile.

BTW: target "useradd seems irrelevant, I'll check this and remove it"

Hope this few words may help you to start with Docker containers.

**Comment by patrickjquinn** [comment]
Sorry sorry, I meant attach a shell, not ssh :)

**Comment by extrowerk** [comment]
I am really time-constrained lately because IRL job, whole word in flames, release at friday, after that i will look into it again.
Thanks!

**Comment by karawitan** [comment]
Building rust on bb10 needs ninja first .. might be quite easy.
2 dirty patches already ready here
https://github.com/berryamin/ninja/tree/release
